middleware.go

```diff
@@ -90,10 +92,23 @@ func (b *LocalDialBreaker) Trip() {
 // ProxyMiddleware adds a Proxy to the given transport. This implementation
 // uses the http.ProxyFromEnvironment.
 func ProxyMiddleware(transport *http.Transport) *http.Transport {
-	transport.Proxy = http.ProxyFromEnvironment
+	transport.Proxy = getProxy
 	return transport
 }
 
+var midLogger = loggo.GetLoggerWithLabels("juju.http.middleware", "http")
+
+func getProxy(req *http.Request) (*url.URL, error) {
+	// Get proxy config new for each client.  Go will cache the proxy
+	// settings for a process, this is a problem for long running programs.
+	// And caused changes in proxy settings via model-config not to
+	// be used.
+	cfg := httpproxy.FromEnvironment()
+	midLogger.Tracef("proxy config http(%s), https(%s), no-proxy(%s)",
+		cfg.HTTPProxy, cfg.HTTPSProxy, cfg.NoProxy)
```

Comment:
I thought `http.ProxyFromEnvironment` was the thing that was doing the caching, which means that this is still using the cached values.

Have we tested that this actually sees the updated values correctly?

Reply:
The dot is in a different place.  The http package versus the httpproxy package. One returns a func the other a Config struct.

I added the logger to ensure we knew what values were being used for the proxy settings with each call. It can be enabled by setting the http log label level to trace, or the new juju.http.middleware log level to trace.